wfe2/verify.go

```diff
@@ -5,246 +5,540 @@ import (
+// HTTP request. This prevents a JWS intended for one endpoint being replayed
+// against a different endpoint. It mutates the provided logEvent to capture any
+// errors.
+func (wfe *WebFrontEndImpl) validPOSTURL(
+	request *http.Request,
+	jws *jose.JSONWebSignature,
+	logEvent *requestEvent) *probs.ProblemDetails {
+	// validPOSTURL is called after parseJWS() which defends against the incorrect
+	// number of signatures.
+	header := jws.Signatures[0].Header
+	extraHeaders := header.ExtraHeaders
+	// Check that there is at least one Extra Header
+	if len(extraHeaders) == 0 {
+		wfe.stats.Inc("Errors.MissingURLinJWS", 1)
+		logEvent.AddError("JWS header parameter 'url' missing")
+		return probs.Malformed("JWS header parameter 'url' required")
+	}
+	// Try to read a 'url' Extra Header as a string
+	headerURL, ok := extraHeaders[jose.HeaderKey("url")].(string)
+	if !ok || len(headerURL) == 0 {
+		wfe.stats.Inc("Errors.MissingURLinJWS", 1)
+		logEvent.AddError("JWS header parameter 'url' missing")
+		return probs.Malformed("JWS header parameter 'url' required")
+	}
+	// Compute the URL we expect to be in the JWS based on the HTTP request
+	expectedURL := url.URL{
+		Scheme: requestProto(request),
+		Host:   request.Host,
+		Path:   request.RequestURI,
+	}
+	// Check that the URL we expect is the one that was found in the signed JWS
+	// header
+	if expectedURL.String() != headerURL {
+		return probs.Malformed(fmt.Sprintf(
+			"JWS header parameter 'url' incorrect. Expected %q got %q",
+			expectedURL.String(), headerURL))
+	}
+	return nil
+}
+
+// parseJWS extracts a JSONWebSignature from an HTTP POST request's body. If
+// there is an error reading the JWS or if it has too few or too many
+// signatures, a problem is returned and the requestEvent is mutated to contain
+// the error.
+func (wfe *WebFrontEndImpl) parseJWS(
+	request *http.Request,
+	logEvent *requestEvent) (*jose.JSONWebSignature, string, *probs.ProblemDetails) {
+	// Verify that the POST request has the expected headers
+	if prob := wfe.validPOSTRequest(request, logEvent); prob != nil {
+		return nil, "", prob
 	}
 
+	// Read the POST request body's bytes. validPOSTRequest has already checked
+	// that the body is non-nil
 	bodyBytes, err := ioutil.ReadAll(request.Body)
 	if err != nil {
 		wfe.stats.Inc("Errors.UnableToReadRequestBody", 1)
 		logEvent.AddError("unable to read request body")
-		return nil, nil, reg, probs.ServerInternal("unable to read request body")
+		return nil, "", probs.ServerInternal("unable to read request body")
 	}
 
+	// Parse the raw JWS JSON to check that the unprotected Header field is not
+	// being used for a key ID or a JWK. This must be done prior to
+	// `jose.parseSigned` since it will strip away these headers.
+	var unprotected struct {
+		Header map[string]string
+	}
+	if err := json.Unmarshal(bodyBytes, &unprotected); err != nil {
+		wfe.stats.Inc("Errors.JWSParseError", 1)
+		logEvent.AddError("Parse error reading JWS from POST body")
```

Comment:
See https://github.com/letsencrypt/boulder/issues/2939; I've been holding off on asking you to remove duplicated log messages from all the code you touch, since this is already a big PR. But since this is new code, we should definitely avoid the duplicated log message pattern.